homeassistant/components/pi_hole/config_flow.py

```diff
@@ -55,27 +61,35 @@ async def async_step_init(self, user_input, is_import=False):
             location = user_input[CONF_LOCATION]
             tls = user_input[CONF_SSL]
             verify_tls = user_input[CONF_VERIFY_SSL]
-            api_token = user_input.get(CONF_API_KEY)
             endpoint = f"{host}/{location}"
 
             if await self._async_endpoint_existed(endpoint):
                 return self.async_abort(reason="already_configured")
 
             try:
-                await self._async_try_connect(
-                    host, location, tls, verify_tls, api_token
-                )
-                return self.async_create_entry(
-                    title=name,
-                    data={
-                        CONF_HOST: host,
-                        CONF_NAME: name,
-                        CONF_LOCATION: location,
-                        CONF_SSL: tls,
-                        CONF_VERIFY_SSL: verify_tls,
-                        CONF_API_KEY: api_token,
-                    },
-                )
+                await self._async_try_connect(host, location, tls, verify_tls)
+                self._config = {
+                    CONF_HOST: host,
+                    CONF_NAME: name,
+                    CONF_LOCATION: location,
+                    CONF_SSL: tls,
+                    CONF_VERIFY_SSL: verify_tls,
+                }
+                if is_import:
+                    api_key = user_input.get(CONF_API_KEY)
+                    return self.async_create_entry(
+                        title=name,
+                        data={
+                            **self._config,
+                            CONF_STATISTICS_ONLY: api_key is None,
+                            CONF_API_KEY: api_key,
+                        },
+                    )
+                statistics_only = user_input[CONF_STATISTICS_ONLY]
+                self._config[CONF_STATISTICS_ONLY] = statistics_only
+                if statistics_only:
+                    return self.async_create_entry(title=name, data=self._config)
+                return await self.async_step_api_key()
             except HoleError as ex:
```

Comment:
Which method can cause the `HoleError`, we should try to wrap only the relevant code that does this. You can either use `try…except…else` to run code in case it didn't raise, or you rewrite this method to be `if user_input is None: return self._show_user_form(defaults, errors)` and so return each time you know you don't need to execute the rest of the function anymore.